### ISSUE TITLE
Merge DottyPlugin in sbt

### ISSUE THREAD

**Issue by adpi2** [sbt/sbt#6080]
The `sbt-dotty` plugin sources can be found here: https://github.com/lampepfl/dotty/tree/master/sbt-dotty/src/dotty/tools/sbtplugin

It is composed of two independent parts:
- `DottyIDEPlugin`: install and configure vscode Dotty extension, start vscode
- `DottyPlugin`: manage Scala instance for Scala 3, configure `compile` and `doc`

This issue is scoped to merging the `DottyPlugin` logic into `sbt`. The `DottyIDEPlugin` will not be merged.

I give below a detailed description of the plugin and I propose an equivalent implementation in sbt.

Most of the time, the logic can be moved to either the default setting value or to the upstream libraries (librarymanagement, zinc).

# Description

## Scala 3 Instances

| Key | Explanation | Proposed Implementation |
|-|-|-|
| `isDotty` | new boolean setting | There already is `ScalaInstance.isDotty` in sbt but the setting is easier to use and it will be very useful during the migration to Scala 3.<br> `isDotty` will be renamed `isScala3` |
| `managedScalaInstance` | `false` if `isDotty`.<br>The Scala instances are fetched by `sbt-dotty` | Should remain `true`. <br>sbt will manage Scala 3 instances |
| `scalaInstance` | Call `dependencyResolution.value.update` to fetch the Scala 3 compiler and library, as well as the Scala 2.13 library | Update `ScalaArtifacts` in `librarymanagement` to manage Scala 3 artifacts.<br>**Should we support Dotty versions (0.xxx)?** |
| `libraryDependencies` | Add the `scala3-library` dependency. | Update `Defaults.autoLibraryDependency` to add `scala3-library` instead of `scala-library`<br>(I guess `scala-library_2.13` is added transitively by the `scala3-library` dependency) |
| `scalaBinaryVersion` | Add some logic for RC versions:<br>The binary version of `0.27.0-RC1` is `0.27`<br>The binary version of `3.0.0-M1-bin-20201027-b5a1715-NIGHTLY` is `3.0.0-M1` | Update the default `scalaBinaryVersion` logic if necessary.<br>(Will have RC versions after for `3.0.0`?) |

## Zinc Config

| Key | Explanation | Proposed Implementation |
|-|-|-|
| `incOptions` | Add a `TastyFileManager` as an external hook of the `incOptions` so that the `.tasty` files are pruned  | Handle in Zinc |
| `scalaCompilerBridgeBinaryJar` | Call `dependencyResolution.value.update` to fetch the `sbt-bridge` jars. | Merge `sbt-bridge` in Zinc.<br>Update `ZincComponentCompiler.getDefaultBridgeModule`. |

## Compiler classpath and bootclasspath

Quoting `sbt-dotty` comment:

> We want:
>
> 1. Nothing but the Java standard library on the _JVM_ bootclasspath (starting with Java 9 we cannot inspect it so we don't have a choice)
>
> 2. scala-library, dotty-library, dotty-compiler and its dependencies on the _JVM_classpath, because we need all of those to actually run the compiler.
>    NOTE: All of those should have the *same version* (equal to scalaVersion for everything but scala-library). (Complication: because dottydoc is a separate artifact with its own dependencies, running it requires putting extra dependencies on the _JVM_ classpath)
>
> 3. scala-library, dotty-library on the _compiler_ bootclasspath or classpath (the only difference between them is that the compiler bootclasspath has higher priority, but that should never make a difference in a sane environment).
>    NOTE: the versions of {scala,dotty}-library used here do not necessarily match the one used in 2. because a dependency of the current project might require a more recent standard library version, this is OK
>    TODO: ... but if macros are used we might be forced to use the same versions in the JVM and compiler classpaths to avoid problems, this needs to be investigated.
>
> 4. every other dependency of the user project on the _compiler_ classpath.
>
> By default, zinc will put on the compiler bootclasspath the scala-library used on the JVM classpath, even if the current project transitively depends on a newer scala-library (this works because Scala 2 guarantees forward- and backward- binary compatibility, but we don't necessarily want to keep doing that in Scala 3).
> So for the moment, let's just put nothing at all on the compiler bootclasspath, and instead let sbt dependency management choose which scala-library and dotty-library to put on the compiler classpath.
> **Maybe eventually we should just remove the compiler bootclasspath since it's a source of complication with only dubious benefits.**

My understanding is that `sbt-dotty` is hacking the `classpathOptions` to prevent Zinc from adding the wrong `scala-library-3.0` on the `bootClasspath`.

I propose to let the `classpathOptions` unchanged and add some logic in Zinc to resolve the correct scala library.

| Key | Explanation | Proposed Implementation |
|-|-|-|
| `classpathOptions` | If `isDotty` then `autoBoot` and `filterLibrary` are set to false. | Do not change the `classpathOptions`.<br>Instead Zinc should add the correct scala library |
| `console / classpathOptions` | The `console / classpathOptions` must not be affected by the above `classpathOptions` override. | Should not be needed |
| `scalacOptions` | If `isDotty` and Java 8, the compiler bootclasspath must contain the JVM bootclasspath<br>to prevent Zinc from breaking. | Handle in Zinc |
| `scalaModuleInfo` | If `isDotty` it deactivates the override of `scala-library` and `scala-compiler` versions. | Backport to the default `scalaModuleInfo`:<br>`overrideScalaVersion = isScala3.value` |

## doc task

| Key | Explanation | Proposed Implementaion |
|-|-|-|
| `useScala3doc` | New boolean setting to use the new `scala3doc` rather than `dotty-doc`.<br>Default to false. | **Will `scala3doc` become the only doc tool for `Scala 3`?**<br>If so this setting will not be added. |
| `resolvers` | If `useScala3doc` add the `jcenterRepo` | **Will `scala3doc` be published to Maven Central?**<br>If so this is not needed. |
| `scala3docOptions` | New `Seq[String]` setting | I dont think we need a new setting for this:<br>We can use `doc / scalacOptions` instead |
| `Compile / doc / scalacOptions` | If `useScala3doc` add `scala3docOptions` with the `"--+DOC+"` prefix There are a few questions that remain.| The `"--+DOC+"` prefix could be added by a utility method:<br>`Compile / doc / scalacOptions ++= Seq("-foo", "-bar").map(_.withDocPrefix)` |
| `doc / scalaInstance` | Depending on `isDotty` and `useScala3doc` it loads `scala3doc`, `dotty-doc` or the default `doc / scalaInstance` | When `scala3doc` replace `dotty-doc` then backport the logic to the default `doc / scalaInstance` |
| `<Compile \| Test> / doc / sources` | If `isDotty` then compile and add the `.tasty` files | Backport the logic to the default `doc / sources` in `configSettings` |
| `<Compile \| Test> / doc / scalacOptions` | If `isDotty` add `"-project ${name.value}"` and `"-from-tasty"` | Backport the logic to the default `doc / sources` in `configSettings` |

## `withDottyCompat`

Quoting `sbt-dotty` comment:

> If this ModuleID cross-version is a Dotty version, replace it by the Scala 2.x version that the Dotty version is retro-compatible with, otherwise do nothing.
>
> This setting is useful when your build contains dependencies that have only been published with Scala 2.x, if you have:
> ```scala
> libraryDependencies += "a" %% "b" % "c"
> ```
> you can replace it by:
> ```scala
> libraryDependencies += ("a" %% "b" % "c").withDottyCompat(scalaVersion.value)
> ```
> This will have no effect when compiling with Scala 2.x, but when compiling with Dotty this will change the cross-version to a Scala 2.x one.
> This works because Dotty is currently retro-compatible with Scala 2.x.
>
> NOTE: As a special-case, the cross-version of scala3-library and scala3-compiler
>  will never be rewritten because we know that they're Scala 3 only.
>  This makes it possible to do something like:
>  ```scala
>  libraryDependencies ~= (_.map(_.withDottyCompat(scalaVersion.value)))
>  ```

`withDottyCompat` only deals with Scala 3 module depending on Scala 2.13 artifact. This was, so far, the only useful pattern of cross binary compatibility.

But we can believe that such patterns will diversified. For instance, it will  soon be possible to have a dependency to a Scala 3 artifact in a Scala 2.13.4 module.

So I propose to generalize `withDottyCompat` with the following: 

```
def withCompat(mapping: (String, String)*): ModuleID
```
so that we can do
```scala
libraryDependencies += ("a" %% "b" % "c").withCompat("x" -> "y")
```
meaning that if the current binary version is "x" we used "y" instead.

The special-case described above won't be supported which is okay since it will be less and less useful.

## Scala-js

The Scala.js specific logic will not be merged into sbt.
Instead it is moved to `sbt-scalajs` (https://github.com/scala-js/scala-js/pull/4280).

# Roadmap

### 1. Merge `sbt-bridge` in Zinc
  - Update `ZincComponentCompiler.getDefaultBridgeModule`
### 4. Manage Scala 3 instances:
  - Add `isScala3` setting
  - Update `ScalaArtifacts` in `librarymanagement`
  - Update `Defaults.autoLibraryDependency`
  - Update `scalaBinaryVersion` if necessary
### 6. Add support for `scala3doc`:
  - Update the `doc / scalaInstance`
  - Update the `doc / sources`
  - Update the `doc / scalacOptions` 

# Unsettled questions

A few questions on which I would like to have someone else's opinion.

### 1. Should we support Dotty versions (0.xxx)?

Pros:
- As of today, library authors want to be compatible with both `0.27.0` and `3.0.0-M1` (see [blogpost](https://gist.github.com/djspiewak/b8f2b4547951442102488964bc351cf9))

Cons:
- Will soon be unecessary complexity: Dotty versions won't be used anymore
- `sbt-dotty` can still be used to handle Dotty versions

IMO sbt should not support `0.27.0` and earlier versions.

@sjrd, @eed3si9n and others: What is your opinion?

### 2. Does it work to have the scala-library and scala3-library on the compiler bootclasspath?

I guess yes and I also think it is easier given that sbt does it this way for Scala 2.

@smarter Do you confirm?

### 3. Will `scala3doc` become the only doc tool for `Scala 3`? Will it be published to Maven Central?

@abgruszecki Can you answer these questions?

### 4. Do you agree with the generalization of `withDottyCompat` as described above?

@sjrd, @eed3si9n, @smarter what do you think?

**Comment by smarter** [comment]
> isDotty

I think we should stop using the name "dotty" outside of the compiler internals, so if we keep it, it should be `isScala3`, but now that we're using >= 3 version numbers, this could also be replaced by something like `VersionNumber(scalaVersion.value).matchesSemVer(SemanticSelector(">= 3.0.0-M1"))` using https://www.scala-sbt.org/1.x/docs/sbt-1.2-Release-Notes.html#Semantic+Version+selector+API which is longer but more flexible.

> Should we support Dotty versions (0.xxx)?

No need, people can keep sbt-dotty while they keep using old dotty versions.

> Update the default scalaBinaryVersion logic if necessary.

The exact thing to do here will be determined by https://github.com/lampepfl/dotty/issues/10244

> My understanding is that sbt-dotty is hacking the classpathOptions to prevent Zinc from adding the wrong scala-library-3.0 on the bootClasspath.

So this used to be the case but I think it should work correctly as of recent zinc versions as I added support for passing multiple library jars to `ScalaInstance`, the remaining reason for customizing this stuff was an attempt at allowing a project to set its scalaVersion to 3.0.0 but to use the standard library from Scala 3.0.1 if it transitively depends on it, but this won't be needed if we preserver backward and forward binary compatibility in the 3.0.x time frame which seems to be where we're headed in https://github.com/lampepfl/dotty/issues/10244

> Will scala3doc become the only doc tool for Scala 3?

Yes.

> Will scala3doc be published to Maven Central?

https://github.com/lampepfl/scala3doc/issues/226

> libraryDependencies += ("a" %% "b" % "c").withCompat("x" -> "y")

I don't know what the right API is here but I think that one is too easy to misuse, there's only two compatibility modes we have currently (2.13 to 3 and 3 to 2.13), so we shouldn't accept arbitrary inputs, maybe `withCompat` could take an enum instead?

**Comment by eed3si9n** [comment]
> 1. Should we support Dotty versions (0.xxx)?

+1 on dropping 0.27.0 support.

> 4. Do you agree with the generalization of withDottyCompat as described above?

In https://github.com/lampepfl/dotty/pull/9974#discussion_r502463586 @julienrf had a good insight that we already have `cross` that does ModuleID conversion. In other words, we could come up with some `cross(...)` rules that would change the interpretation of `%`. https://www.scala-sbt.org/1.x/docs/Cross-Build.html#More+about+using+cross-built+libraries

> This uses a custom function to determine the Scala version to use based on the binary Scala version:

```scala
("a" % "b" % "1.0") cross CrossVersion.binaryMapped {
  case "2.10" => "2.10.0" // useful if a%b was released with the old style
  case x => x
}
```

I wonder if we could do something like:

```scala
("a" % "b" % "1.0").cross(CrossVersion.for3_use2_13)

// equivalent to
("a" % "b" % "1.0") cross CrossVersion.binaryMapped {
  case "3.0" => "2.13" // use _2.13 only for Scala 3.0
  case x => x
}
```

**Comment by sjrd** [comment]
I agree with everything that was said.

For `withDottyCompat`, we should definitely stick to something that is limited to "use 2.13 for 3" and "use 3 for 2.13". @eed3si9n's proposal seems reasonable.

**Comment by abgruszecki** [comment]
Reg. Scala3doc: in The End, we should:
1) not need `useScala3doc` as there will be only one doctool for Scala3
2) not need to add resolvers as the dependencies of Scala3doc should be published to Maven Central
3) not need to add the magick prefix to Scala3doc-specific options

No. 2) and 3) need updates upstream in Scala3doc, though. In particular reg. 3), right now the magick prefix should only be added to Scala3doc-specific options, which is the reason why for the time being we have a separate setting key for these options.

What's the timeline for these changes in SBT? As long as we only do them after Scala3doc was updated, we're golden in this particular area.

**Comment by eed3si9n** [comment]
> What's the timeline for these changes in SBT? 

If we want sbt 1.5.0 to be out definitely by February, 2021, I think we should aim for RC-1 by Christmas, 2020.

**Comment by smarter** [comment]
I don't fully get the sbt versioning policy, but couldn't these changes go into a 1.4.x release?

**Comment by sjrd** [comment]
If there is at least one new public member anywhere, such as a setting, it won't be forward binary compatible, and that means 1.5.0.

**Comment by smarter** [comment]
but sbt sometimes introduces new settings in patch versions, e.g. https://github.com/sbt/sbt/releases/tag/v1.3.2 says:
> Adds a new setting called includePluginResolvers (default false).

**Comment by adpi2** [comment]
We don't really need to add new public members:
- `isScala3` and `withCompat` can be delayed to `1.5`
- we can advocate for alternative solutions: `matchesSemVer`, `binaryMapped`
- the new implementation could be compatible with `sbt-dotty` (maybe not the `TastyFileManager` though)

So technically it can be shipped into `1.4.x`.

However, if we ship it in `1.5.0` we will have `RC` versions to test it and we will eventually be more confident.

**Comment by eed3si9n** [comment]
> However, if we ship it in `1.5.0` we will have RC versions to test it and we will eventually be more confident.

Yea it's that basically. Whether something 1.4.x or 1.5.0 is determined more by whether it's likely to break the existing build.
New features tend to include regressions, so we put them in 1.X.0 so we can go through release candidates. I'd be happy to release M1, M2 etc as soon as it merges in addition.

**Comment by xuwei-k** [comment]
Should we auto add `"-Ytasty-reader"` option if sandwich?

https://github.com/sbt/sbt/pull/6121
https://github.com/scala/scala/commit/3743cad6cef6dd8a13c870a66104225d456942c1

**Comment by smarter** [comment]
No, the point of making it a flag is that it's experimental so we want people to set it explicitly if they want to try it

**Comment by smarter** [comment]
One thing that isn't handled by sbt-dotty but should be handled once dotty support is integrated in sbt: the `semanticdbEnabled` setting which is built into sbt (for dotty, this should just enable the `-Ysemanticdb` flag)

**Comment by adpi2** [comment]
@smarter that is already handled by sbt

**Comment by adpi2** [comment]
About the classpath configuration @smarter said
> So this used to be the case but I think it should work correctly as of recent zinc versions as I added support for passing multiple library jars to ScalaInstance, the remaining reason for customizing this stuff was an attempt at allowing a project to set its scalaVersion to 3.0.0 but to use the standard library from Scala 3.0.1 if it transitively depends on it, but this won't be needed if we preserver backward and forward binary compatibility in the 3.0.x time frame which seems to be where we're headed in lampepfl/dotty#10244

I confirm that the default `-bootclasspath` and `-classpath` configuration is working well with Scala 3.

The `-bootclasspath` option contains:
- `/home/piquerez/.cache/coursier/jvm/adopt@1.8.0-252/jre/lib/resources.jar`
- `/home/piquerez/.cache/coursier/jvm/adopt@1.8.0-252/jre/lib/rt.jar`
- `/home/piquerez/.cache/coursier/jvm/adopt@1.8.0-252/jre/lib/sunrsasign.jar`
- `/home/piquerez/.cache/coursier/jvm/adopt@1.8.0-252/jre/lib/jsse.jar`
- `/home/piquerez/.cache/coursier/jvm/adopt@1.8.0-252/jre/lib/jce.jar`
- `/home/piquerez/.cache/coursier/jvm/adopt@1.8.0-252/jre/lib/charsets.jar`
- `/home/piquerez/.cache/coursier/jvm/adopt@1.8.0-252/jre/lib/jfr.jar`
- `/home/piquerez/.cache/coursier/jvm/adopt@1.8.0-252/jre/classes`
`/home/piquerez/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3.0.0-M1/3.0.0-M1/scala3-library_3.0.0-M1-3.0.0-M1.jar`
- `/home/piquerez/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.3/scala-library-2.13.3.jar`

And `-classpath` contains (I don't have any dependencies):
- `/example/target/scala-3.0.0-M1/classes`

**Comment by adpi2** [comment]
I share with you a status of my progress on this task:

- [x] **Scala 3 instance class loader**
Unlike the Scala 2 instance, the Scala 3 instance needs the `org.scala-sbt:compiler-interface` in its class loader.
It has been added in sbt-dotty for compatibility with sbt `1.3.x` and `1.4.x`
It has been added by default for all Scala instances in https://github.com/sbt/sbt/pull/6199 (sbt `1.5.x`)
Also it has been added in Mill (https://github.com/lihaoyi/mill/pull/1019)
- [x] **Scala 3 compiler bridge**
The Scala 3 compiler bridge has been upgraded to Zinc 1.4.3 in https://github.com/lampepfl/dotty/pull/10607
It still provides an implementation of the old interface for compatibility with sbt 1.3.x and Mill. Once Mill has upgraded to Zinc 1.4.3 we will be able to remove the old `xsbt.CompilerInterface` implementation.
- [x] **Add `TastyFileManager` logic in Zinc**: https://github.com/sbt/zinc/pull/956
- [x] **Manage Scala 3 compiler instance in sbt**: In progress in https://github.com/sbt/librarymanagement/pull/350 and https://github.com/sbt/sbt/pull/6206
- [x] **Manage Scala 3 doc instance in sbt**

**Comment by adpi2** [comment]
Also I will fix `consoleProject` for Scala 3.0.0.
It is currently not working because sbt always uses the Scala 2.12 instance instead of the Scala 3.0.0 instance to create the `xsbt.ConsoleInterface`:
https://github.com/sbt/sbt/blob/23fd24c838e6aa68688b01861a4c21f1154f9103/main/src/main/scala/sbt/internal/ConsoleProject.scala#L29-L32

**Comment by nafg** [comment]
> I think we should stop using the name "dotty" outside of the compiler internals, so if we keep it, it should be isScala3, but now that we're using >= 3 version numbers, this could also be replaced by something like VersionNumber(scalaVersion.value).matchesSemVer(SemanticSelector(">= 3.0.0-M1")) using https://www.scala-sbt.org/1.x/docs/sbt-1.2-Release-Notes.html#Semantic+Version+selector+API which is longer but more flexible.

Also that could (IMO should) be made shorter (but I guess this is not the place to get into details)

**Comment by adpi2** [comment]
@eed3si9n 

I need your opinion about the management of the `scala3doc` artifact.

In Scala 2, the doc compiler is provided by the `scala-compiler` artifact so we can reuse the `<project> / scalaInstance` in the `doc` task.

In Scala 3 the doc compiler is provided by a distinct `scala3doc` artifact. The `<project> / scalaInstance` does not contain it in sbt-dotty nor in #6206. That's why `sbt-dotty` creates a `doc / scalaInstance` by fetching `scala3doc` explicitly. That does the job. But it does not integrate very well with the `tool` configuration logic because `scala3doc` does not appear in `allDependencies` or in the  `ivyModule.directDependencies`.

Alternatively I can simply add `scala3doc` in the `ScalaArtifacts.toolDependencies` method. It is a simple solution but it forces every Scala 3 project to fetch `scala3doc` even if they don't use the `doc` task.

What's your favorite solution?
1- Fetch `scala3doc` in `doc / scalaInstance` under the hood.
2- Add `scala3doc` in `ScalaArtifacts.toolDependencies` (as in https://github.com/sbt/librarymanagement/pull/359).

@smarter suggested that maybe we should stop using the `tool` configuration. I am not really convinced and if we go that way I would rather do it in a separate project.

**Comment by eed3si9n** [comment]
@adpi2 Thanks for the ping. I think we need to explore what scala3doc _is_ before we can make the decision.
In Scala 2 the doc compiler was part of the compiler, so we didn't have to worry about (additional) library dependencies.

Now Scala 3 will be based on Kotlin's documentation engine, Flexmark, and Jackson?! https://github.com/lampepfl/dotty/blob/854f58331b5ebd86ffb9326894a4870cb3ade5e7/project/Build.scala#L1548-L1560

```scala
          libraryDependencies ++= Seq(
            "org.jetbrains.dokka" % "dokka-core" % dokkaVersion,
            "org.jetbrains.dokka" % "dokka-base" % dokkaVersion,
            "org.jetbrains.kotlinx" % "kotlinx-html-jvm" % "0.7.2", // Needs update when dokka version changes
            "com.vladsch.flexmark" % "flexmark-all" % "0.42.12",
            "nl.big-o" % "liqp" % "0.6.7",
            "org.jsoup" % "jsoup" % "1.13.1", // Needed to process .html files for static site
            "args4j" % "args4j" % "2.33",
            Dependencies.`jackson-dataformat-yaml`,

            "org.jetbrains.dokka" % "dokka-test-api" % dokkaVersion % "test",
            "com.novocode" % "junit-interface" % "0.11" % "test",
          ),
```

The main motivation for using `tool` configuration is so the resolution goes through the normal library management channel (necessary since some companies would go out to the Internet via Artifactory etc) and to do the dependency resolution in a single shot.

I don't think it's bad to grab all the necessary JAR in the very first use of a Scala 3 version, but with this kitchen sink of additional dependencies, we probably need to be more careful about _how_ ScalaInstance is used to set up the ClassLoader for _regular_ compilation.

**Comment by adpi2** [comment]
Okay so maybe we should use the same `ScalaInstance` for `doc` but we should also be able to distinguish between the `loaderCompilerOnly`, that does not contain the `scala3doc` and all its dependencies, and the `loader` that contains all the jars.

So adding an extra layer in the ScalaInstance class loader...

It would not make any difference for Scala 2 but it will give extra safety in Scala 3 for regular compilation.

**Comment by eed3si9n** [comment]
Here's the current list of JARs after locally fixing the Dotty build since it's currently not hooked up to anything:

```
sbt:scala3> show scala3doc/Compile/externalDependencyClasspath
[info] * Attributed(.../https/jcenter.bintray.com/org/jetbrains/dokka/dokka-core/1.4.10.2/dokka-core-1.4.10.2.jar)
[info] * Attributed(.../https/jcenter.bintray.com/org/jetbrains/dokka/dokka-base/1.4.10.2/dokka-base-1.4.10.2.jar)
[info] * Attributed(.../https/jcenter.bintray.com/org/jetbrains/kotlinx/kotlinx-html-jvm/0.7.2/kotlinx-html-jvm-0.7.2.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-all/0.42.12/flexmark-all-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/nl/big-o/liqp/0.6.7/liqp-0.6.7.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jsoup/jsoup/1.13.1/jsoup-1.13.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/args4j/args4j/2.33/args4j-2.33.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/fasterxml/jackson/dataformat/jackson-dataformat-yaml/2.9.8/jackson-dataformat-yaml-2.9.8.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/scala-lang/modules/scala-asm/7.3.1-scala-1/scala-asm-7.3.1-scala-1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/scala-sbt/compiler-interface/1.3.5/compiler-interface-1.3.5.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jline/jline-reader/3.15.0/jline-reader-3.15.0.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jline/jline-terminal/3.15.0/jline-terminal-3.15.0.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jline/jline-terminal-jna/3.15.0/jline-terminal-jna-3.15.0.jar)
[info] * Attributed(.../https/jcenter.bintray.com/org/jetbrains/markdown/0.1.45/markdown-0.1.45.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-stdlib-jdk8/1.4.10/kotlin-stdlib-jdk8-1.4.10.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-reflect/1.4.10/kotlin-reflect-1.4.10.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/fasterxml/jackson/module/jackson-module-kotlin/2.11.1/jackson-module-kotlin-2.11.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/fasterxml/jackson/dataformat/jackson-dataformat-xml/2.11.1/jackson-dataformat-xml-2.11.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines-core-jvm/1.3.9/kotlinx-coroutines-core-jvm-1.3.9.jar)
[info] * Attributed(.../https/jcenter.bintray.com/org/jetbrains/dokka/dokka-analysis/1.4.10.2/dokka-analysis-1.4.10.2.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-stdlib/1.4.10/kotlin-stdlib-1.4.10.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-stdlib-common/1.4.10/kotlin-stdlib-common-1.4.10.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark/0.42.12/flexmark-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-abbreviation/0.42.12/flexmark-ext-abbreviation-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-admonition/0.42.12/flexmark-ext-admonition-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-anchorlink/0.42.12/flexmark-ext-anchorlink-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-aside/0.42.12/flexmark-ext-aside-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-attributes/0.42.12/flexmark-ext-attributes-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-autolink/0.42.12/flexmark-ext-autolink-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-definition/0.42.12/flexmark-ext-definition-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-emoji/0.42.12/flexmark-ext-emoji-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-enumerated-reference/0.42.12/flexmark-ext-enumerated-reference-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-escaped-character/0.42.12/flexmark-ext-escaped-character-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-footnotes/0.42.12/flexmark-ext-footnotes-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-gfm-issues/0.42.12/flexmark-ext-gfm-issues-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-gfm-strikethrough/0.42.12/flexmark-ext-gfm-strikethrough-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-gfm-tables/0.42.12/flexmark-ext-gfm-tables-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-gfm-tasklist/0.42.12/flexmark-ext-gfm-tasklist-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-gfm-users/0.42.12/flexmark-ext-gfm-users-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-gitlab/0.42.12/flexmark-ext-gitlab-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-jekyll-front-matter/0.42.12/flexmark-ext-jekyll-front-matter-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-jekyll-tag/0.42.12/flexmark-ext-jekyll-tag-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-media-tags/0.42.12/flexmark-ext-media-tags-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-macros/0.42.12/flexmark-ext-macros-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-ins/0.42.12/flexmark-ext-ins-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-xwiki-macros/0.42.12/flexmark-ext-xwiki-macros-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-superscript/0.42.12/flexmark-ext-superscript-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-tables/0.42.12/flexmark-ext-tables-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-toc/0.42.12/flexmark-ext-toc-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-typographic/0.42.12/flexmark-ext-typographic-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-wikilink/0.42.12/flexmark-ext-wikilink-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-yaml-front-matter/0.42.12/flexmark-ext-yaml-front-matter-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-ext-youtube-embedded/0.42.12/flexmark-ext-youtube-embedded-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-formatter/0.42.12/flexmark-formatter-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-html-parser/0.42.12/flexmark-html-parser-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-jira-converter/0.42.12/flexmark-jira-converter-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-pdf-converter/0.42.12/flexmark-pdf-converter-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-profile-pegdown/0.42.12/flexmark-profile-pegdown-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-util/0.42.12/flexmark-util-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/vladsch/flexmark/flexmark-youtrack-converter/0.42.12/flexmark-youtrack-converter-0.42.12.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/antlr/antlr/3.5.1/antlr-3.5.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-annotations/2.11.1/jackson-annotations-2.11.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-core/2.11.1/jackson-core-2.11.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-databind/2.11.1/jackson-databind-2.11.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/yaml/snakeyaml/1.23/snakeyaml-1.23.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.4/scala-library-2.13.4.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/google/protobuf/protobuf-java/3.7.0/protobuf-java-3.7.0.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/scala-sbt/util-interface/1.3.0/util-interface-1.3.0.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/net/java/dev/jna/jna/5.3.1/jna-5.3.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-stdlib-jdk7/1.4.10/kotlin-stdlib-jdk7-1.4.10.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/fasterxml/jackson/module/jackson-module-jaxb-annotations/2.11.1/jackson-module-jaxb-annotations-2.11.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/codehaus/woodstox/stax2-api/4.2.1/stax2-api-4.2.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/fasterxml/woodstox/woodstox-core/6.2.1/woodstox-core-6.2.1.jar)
[info] * Attributed(.../https/jcenter.bintray.com/org/jetbrains/dokka/kotlin-analysis-intellij/1.4.10.2/kotlin-analysis-intellij-1.4.10.2.jar)
[info] * Attributed(.../https/jcenter.bintray.com/org/jetbrains/dokka/kotlin-analysis-compiler/1.4.10.2/kotlin-analysis-compiler-1.4.10.2.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/jetbrains/annotations/13.0/annotations-13.0.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/nibor/autolink/autolink/0.6.0/autolink-0.6.0.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/openhtmltopdf/openhtmltopdf-core/0.0.1-RC15/openhtmltopdf-core-0.0.1-RC15.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/openhtmltopdf/openhtmltopdf-pdfbox/0.0.1-RC15/openhtmltopdf-pdfbox-0.0.1-RC15.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/openhtmltopdf/openhtmltopdf-rtl-support/0.0.1-RC15/openhtmltopdf-rtl-support-0.0.1-RC15.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/openhtmltopdf/openhtmltopdf-jsoup-dom-converter/0.0.1-RC15/openhtmltopdf-jsoup-dom-converter-0.0.1-RC15.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/antlr/antlr-runtime/3.5.1/antlr-runtime-3.5.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/antlr/ST4/4.0.7/ST4-4.0.7.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/jakarta/xml/bind/jakarta.xml.bind-api/2.3.2/jakarta.xml.bind-api-2.3.2.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/jakarta/activation/jakarta.activation-api/1.2.1/jakarta.activation-api-1.2.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/apache/pdfbox/pdfbox/2.0.11/pdfbox-2.0.11.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/apache/pdfbox/xmpbox/2.0.11/xmpbox-2.0.11.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/de/rototor/pdfbox/graphics2d/0.15/graphics2d-0.15.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/com/ibm/icu/icu4j/59.1/icu4j-59.1.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/org/apache/pdfbox/fontbox/2.0.11/fontbox-2.0.11.jar)
[info] * Attributed(.../https/repo1.maven.org/maven2/commons-logging/commons-logging/1.2/commons-logging-1.2.jar)
```

@romanowski says the Kotlin dependency would likely be removed in the future?

**Comment by romanowski** [comment]
Sorry for being late to the party :)

We would need 2-3 weeks to get rid of Kotlin deps if this is really necessary. We've started as a plugin for dokka but we gradually moving away into being a standalone thing and we would need to accelarate that process.

This leaves other dependencies that are core to the scala3doc:
 - flexmark: for dealing with markdown in API comments nad static pages)
 - liqp: for building static site
 - jsoup: for building static site (precisely to link to API pages from static site that may be defined in .html)
 - Jackson: we are using it for some things but we should be able to remove that if we really need

So we can clear some of our dependencies but it will be not possible to get it e.g. one external jar.

**Comment by eed3si9n** [comment]
If it's meant to be a standalone app, rather than a library, then one way to at least minimize the download would be to turn it into a fat JAR? 
I just suggested that also so we won't have JCenter artifacts - https://github.com/lampepfl/dotty/issues/10904

**Comment by romanowski** [comment]
The more I think about the proper solution the more I see that compiler and doctool should have separated. 

Scala3doc is loosely coupled with a compiler so `compilation` and `doc` may use different `scalaVersion` (or to be precise `scalaVersion` may be different than `scala3doc` one).

The only relation that we need to have with a compiler is using compatible version of TASTY. As long as we do not expose it as first class citizen in sbt we can keep same scalaVersion for both `doc` and `compile`

**Comment by romanowski** [comment]
> If it's meant to be a standalone app, rather than a library, then one way to at least minimize the download would be to turn it into a fat JAR?

I am not a big fan of fat jars so I see it as a last resort. We are planning to expose parts of scala3doc to allow extensions but we will turn those into API jars anyway.

> I just suggested that also so we won't have JCenter artifacts - lampepfl/dotty#10904

Jcenter was needed for dokka but it was recently disted to maven central so we should be able to remove it.

**Comment by eed3si9n** [comment]
> The more I think about the proper solution the more I see that compiler and doctool should have separated.

When a new Scala version comes out today, the user can say `++2.13.4!` and cross build (including `doc`).
In the world where doctool is separated, do we basically use the same doctool throughout the `3.x` lifecycle (ignoring `3.T`) since TASTY compatibility would be maintained?

**Comment by smarter** [comment]
> Scala3doc is loosely coupled with a compiler so compilation and doc may use different scalaVersion (or to be precise scalaVersion may be different than scala3doc one).

I don't think that's a good idea, the version of the compiler used by scala3doc needs to be as recent as the version used for compilation since minor releases (3.1, 3.2, ...) will bump the tasty minor version. And if sbt has to load two compiler instances then documentation generation will be slower since the JIT will have more work to do, and sbt will have to use more ram.

**Comment by dwijnand** [comment]
Currently every sbt release sets `scalaVersion` to the latest stable version (except when sbt lags or is blocked by something), and then users can upgrade or downgrade it.  Sounds to me like scala3doc should work the same way.  Ideally every Scala 3 version carried with it the "latest and greatest" scala3doc version to use, rather than have that in sbt-dotty / sbt releases - but I can't think of any mechanism we have for that.  Optional dependencies of scala3-compiler?

**Comment by smarter** [comment]
> Ideally every Scala 3 version carried with it the "latest and greatest" scala3doc version to use

The compiler and doctool are released together and use the same version number, there's no need for any complex matching of versions.

**Comment by smarter** [comment]
> Jackson: we are using it for some things but we should be able to remove that if we really need

It looks like you're using it to parse yaml, but you're also using flexmark to do that, are both really needed?

**Comment by smarter** [comment]
> So adding an extra layer in the ScalaInstance class loader...

I think that makes sense, I was also wondering if we should split the repl out of scala3-compiler in its own artifact (it depends on jline, and in the future it might also depend on coursier to provide ivy imports like ammonite), to keep things simple I think we could add just one layer on top of the compiler layer for everything else, and we can put the doc-tool, repl and their dependencies in there.

**Comment by adpi2** [comment]
From the entire list of dependenices of `scala3doc` there are still 2 artifacts that are not available in `repo1.maven.org`:

```
[error] (root / update) sbt.librarymanagement.ResolveException: Error downloading org.jetbrains.kotlinx:kotlinx-html-jvm:0.7.
[error]   Not found
[error]   not found: /home/piquerez/.ivy2/local/org.jetbrains.kotlinx/kotlinx-html-jvm/0.7.2/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-html-jvm/0.7.2/kotlinx-html-jvm-0.7.2.pom
[error] Error downloading org.jetbrains:markdown:0.1.45
[error]   Not found
[error]   not found: /home/piquerez/.ivy2/local/org.jetbrains/markdown/0.1.45/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/jetbrains/markdown/0.1.45/markdown-0.1.45.pom
```

So I have no other choice than temporarily add `jcenterRepo` in the list of resolvers...

**Comment by romanowski** [comment]
@adpi2 @eed3si9n would it be possible to give us a few weeks to migrate away from Dokka? It should make our classpath much shorter, and we will be able to get rid of centre.

> > Jackson: we are using it for some things but we should be able to remove that if we really need

> It looks like you're using it to parse yaml, but you're also using flexmark to do that, are both really needed?

We can take a look at it after migration away from dokka.

**Comment by eed3si9n** [comment]
@romanowski I think it largely depends on when Scala 3 final is coming out and when sbt 1.5.0 is expected to arrive.
Normally the RC cycle takes a few months (every fresh code has bugs) so I was hoping to ship sbt 1.5.0-RC1 this month.
If we're still on schedule with https://www.scala-lang.org/blog/2020/12/15/scala-3-crossing-the-finish-line.html, then I think we should try to get sbt 1.5.0-RC1 out _before_ Scala 3.0.0-RC1 to test all the JAR naming conventions.

So maybe you do have 2 ~ 3 weeks, including this week, but it might be good to just make assembly JAR and push it to Maven Central for the 3.0 version, and do the refactoring afterwards.

Either case, I'll defer the timing decision to @darjutak

**Comment by adpi2** [comment]
@romanowski 
sbt is not really depending on a particular version of `scala3doc`: the `scala3doc` artifacts are fetched at runtime depending on the `scalaVersion`. This should keep being true as long as the Scala 3 compiler and `scala3doc` are released together and use the same version number.

If all goes well, the next `scala3doc` version is `3.0.0-RC1`. So maybe its preferable to release a first milestone of sbt that is compatible with `scala3doc:3.0.0-M3` (so using JCenter by default). Once everybody has jumped to `3.0.0-RC1` we can drop `jcenter` and after a few weeks/months release sbt `1.5.0`.

So it's great if you can migrate away from Dokka and if we can get rid of jcenter. Also I think it will not confilct with the sbt schedule.

**Comment by adpi2** [comment]
The 3 PRs that are related to `scala3doc` are ready for review:
- https://github.com/sbt/librarymanagement/pull/359
- https://github.com/sbt/zinc/pull/960
- https://github.com/sbt/sbt/pull/6234

JCenter is not added by default. The user must set `resolver += Resolver.JCenterRepository` to be able to run the `doc` task on a Scala 3 project. But all other tasks should work fine with or without JCenter. This is meant to be temporary.

Please tell me if you think this is a good strategy or not. It may not be the right timing. If so we can delay the merging of these PRs.

**Comment by adpi2** [comment]
@eed3si9n and others I am trying to implement
```scala
("a" % "b" % "1.0").cross(CrossVersion.for3Use2_13)
```
The `cross` method takes a subtype of `CrossVersion`:
- `Disabled`
- `Binary(prefix: String, suffix: String)`
- `Constant(value: String)`
- `Full(prefix: String, suffix: String)`
- `Patch`

Those subtypes used to be generated but they are not anymore, since https://github.com/sbt/librarymanagement/pull/296.

The `binaryMapped` logic was dropped in sbt `1.0` (see https://github.com/sbt/librarymanagement/pull/117)

So I see 3 possible implementations:
1. Add a `class For3Use2_13() extends CrossVersion with Serializable` and similarly a `For2_13Use3`
2. Reintroduce a `class BinaryMapped(f: String => String) extends CrossVersion` but which cannot be `Serializable` and so escapes `CrossVersionFormats` (idk what  `CrossVersionFormats` is used for)
3. Use `Constant` like `withDottyCompat` does, meaning that `for3Use2_13` takes a `scalaVersion`:
```scala
("a" % "b" % "1.0").cross(CrossVersion.for3Use2_13(scalaVersion.value))
```
What's your take on it?

**Comment by eed3si9n** [comment]
IMO we should go for class For3Use2_13() so ModuleID and UpdateReport etc can be persisted to JSON (plain function cannot be translated into JSON). Scala version info is given separately later for %% so that should not be part of the cross.

**Comment by adpi2** [comment]
Also I guess I should add a `prefix` field for scala.js and Scala native as in `Binary` ([example in ScalaJSPlugin](https://github.com/scala-js/scala-js/blob/master/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSCrossVersion.scala)) and a `suffix`.

So it is:
```scala
class For3Use2_13(prefix: String, suffix: String) extends CrossVersion with Serializable
```

and
```scala
abstract class CrossVersionFunctions {
  def for3Use2_13: CrossVersion = new  For3Use2_13("", "")

  def for3Use2_13With(prefix: String, suffix: String): CrossVersion = new For3Use2_13(prefix, suffix)
}
```

**Comment by adpi2** [comment]
I am getting close to the end of this project.

The only thing that could be seen as missing compared to `sbt-dotty` is the `isDotty` setting that could easily be replaced by a `isScala3.value`. But should we? I don't have a strong opinion. So I give you my thoughts then please tell me what you think.

## `isScala3.value`

### Pros

I think that most people would want it because it is absolutely useful in cross-built projects, to manage the dependencies, the compiler plugins, the scalacOptions among other things.

For instance:
```scala
libraryDependencies ++= {
  if (isScala3.value) Nil
  else Seq(compilerPlugin(("org.typelevel" %% "kind-projector" % kindProjectorVersion).cross(CrossVersion.full)))
},
scalacOptions ++= {
  if (isScala3.value) Seq("-Ykind-projector")
  else Nil
}
```

`isDotty.value` is called 9 times in the cats `build.sbt` file alone: https://github.com/typelevel/cats/blob/master/build.sbt

It is more convenient than writing `scalaVersion.value.startsWith("3")` or any other alternative.


### Cons

But it is not really a setting because it is not something that people might want to override. The following code does not make sense:
```scala
scalaVersion := "3.0.0-M3",
isScala3 := false
```

There are plenty of alternatives:
```scala
scalaVersion.value.startsWith("3.")
ScalaArtifacts.isScala3(scalaVersion.value)
VersionNumber(scalaVersion.value).matchesSemVer(SemanticSelector(">= 3.0.0-M1"))
```

### Should we add a `isScala3` setting and if not which alternative do we advertise (in the migration guide for instance)?

**Comment by nafg** [comment]
Why can't we easily get the Scala version as a Tuple3[Int, Int, Int]?

Also the SemVer thing could have a much nicer DSL.

**Comment by eed3si9n** [comment]
> Should we add a `isScala3` setting and if not which alternative do we advertise (in the migration guide for instance)?

I don't think we should treat Scala 3 any more special than 2.13 etc especially because there might be various conditions the build users may have to make in the future beyond 3.0.

Here's a real-life cross building example that I have on [scopt/scopt](https://github.com/scopt/scopt/blob/8e42dd1de71087e39b0e98dc736cabe79d8e0231/build.sbt#L30-L39):

```scala
    scalacOptions ++= {
      CrossVersion.partialVersion(scalaVersion.value) match {
        case Some((3, _)) =>
          Seq("-source:3.0-migration")
        case Some((2, v)) if v <= 12 =>
          Seq("-Xfuture")
        case _ =>
          Nil
      }
    },
```

Not sure if it's the best, but it works for me.

**Comment by nafg** [comment]
Again, we can make that way easier without special-casing Scala 3

**Comment by adpi2** [comment]
@nafg I think that's exactly the point of @eed3si9n. Or maybe you have something else in mind?

**Comment by nafg** [comment]
Oh I didn't read it closely enough. Not too bad. But there could be a built
in setting that doesn't return Option (unlike CrossVersion.partialVersion
which takes an arbitrary string), like scalaVersionParts. Has the Scala
version format ever not been x.y.z?

**Comment by nafg** [comment]
Oh dotty. Why did they have that weird version scheme. And there's of
course snapshots, milestones, RCs, etc. So it should be a special type, not
a Tuple3. That way you could could say scalaVersionInfo.value.major == 3.

**Comment by adpi2** [comment]
Scala 3 support in sbt is ready to be tested by the users.

I close this issue. For any problem that is related to Scala 3 in sbt please open a new issue and ping me in it.

@eed3si9n can you plan a release for `1.5.0-M1`? A fresh `librarymanagement` version is also needed to have https://github.com/sbt/librarymanagement/pull/361.

**Comment by eed3si9n** [comment]
Sounds good.

**Comment by romanowski** [comment]
@eed3si9n @adpi2 I am way too late for the party but scala3doc does not depend on `dokka` or Kotlin since yesterday. We've just decided that we will drop `scala3doc` in favour of just `scaladoc` and I will change that this week (https://github.com/lampepfl/dotty/issues/11275)
